crates/ruff_python_semantic/src/red_knot/types.rs

```diff
@@ -0,0 +1,705 @@
+    fn own_class_member(self, context: &TypingContext, name: &Name) -> Option<Type> {
+        let class = self.lookup(context);
+
+        let symbols = symbol_table(context.db, class.body_scope);
+        let symbol = symbols.symbol_id_by_name(name)?;
+        let types = context.types(class.body_scope);
+
+        Some(types.symbol_ty(symbol))
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct ClassType {
+    /// Name of the class at definition
+    name: Name,
+
+    /// Types of all class bases
+    bases: Vec<Type>,
+
+    body_scope: ScopeId,
+}
+
+impl ClassType {
+    fn name(&self) -> &str {
+        self.name.as_str()
+    }
+
+    #[allow(unused)]
+    pub(super) fn bases(&self) -> &[Type] {
+        self.bases.as_slice()
+    }
+}
+
+#[newtype_index]
+pub struct LocalUnionTypeId;
+
+impl LocalTypeId for LocalUnionTypeId {
+    type Ty = UnionType;
+
+    fn lookup_local(self, types: &TypeInference) -> &Self::Ty {
+        types.union_ty(self)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct UnionType {
+    // the union type includes values in any of these types
+    elements: FxIndexSet<Type>,
+}
+
+struct UnionTypeBuilder<'a> {
+    elements: FxIndexSet<Type>,
+    context: &'a TypingContext<'a>,
+}
+
+impl<'a> UnionTypeBuilder<'a> {
+    fn new(context: &'a TypingContext<'a>) -> Self {
+        Self {
+            context,
+            elements: FxIndexSet::default(),
+        }
+    }
+
+    fn variant(mut self, ty: Type) -> Self {
+        match ty {
+            Type::Union(union_id) => {
+                let union = union_id.lookup(self.context);
+                self.elements.extend(&union.elements);
+            }
+            _ => {
+                self.elements.insert(ty);
+            }
+        }
+
+        self
+    }
+
+    fn build(self) -> UnionType {
+        UnionType {
+            elements: self.elements,
+        }
+    }
+}
+
+#[newtype_index]
+pub struct LocalIntersectionTypeId;
+
+impl LocalTypeId for LocalIntersectionTypeId {
+    type Ty = IntersectionType;
+
+    fn lookup_local(self, types: &TypeInference) -> &Self::Ty {
+        types.intersection_ty(self)
+    }
+}
+
+// Negation types aren't expressible in annotations, and are most likely to arise from type
+// narrowing along with intersections (e.g. `if not isinstance(...)`), so we represent them
+// directly in intersections rather than as a separate type. This sacrifices some efficiency in the
+// case where a Not appears outside an intersection (unclear when that could even happen, but we'd
+// have to represent it as a single-element intersection if it did) in exchange for better
+// efficiency in the within-intersection case.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct IntersectionType {
+    // the intersection type includes only values in all of these types
+    positive: FxIndexSet<Type>,
+    // the intersection type does not include any value in any of these types
+    negative: FxIndexSet<Type>,
+}
+
```

Comment:
I guess `IntersectionTypeBuilder` (with the intersection-flattening logic) will come along with CFG, where intersection types are actually created.